protos/flyteidl/datacatalog/datacatalog.proto

```diff
@@ -150,6 +157,22 @@ message ListDatasetsResponse {
     string next_token = 2;
 }
 
+message GetOrReserveArtifactRequest {
+    DatasetID dataset_id = 1;
+    string tag_name = 2;
+    google.protobuf.Timestamp expire_at = 3;
+}
+
+enum ReservationStatus {
+    SUCCESS = 0;
```

Comment:
we will have to indicate that a Reservation is already in progress right?

So consider this example

Process 1:
   GetOrReserve(A) -> Reservation(Success)

Process 2:
   GetOrReserve(A) -> Reservation(Failed) - whats the reason for failure?

Also I feel, the timeout should be evaluated on the server side, so how about updating the method

GetOrReserve(A, timeout)

And the Status should be more descriptive
1. ReservationAcquired, ReservationAlreadyInProgress, ReservationFailed, ReservationTimedOut

If the case of ReservationAlreadyInProgress, we should probably also return the holding task.

I would recommend changing
GetOrReserveArtifactRequest to include Metadata - https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/datacatalog/datacatalog.proto#L215

Thus the ReservationStatus is not just an enum, but more like
message ReservationStatus {
   enum State {
      ...
   },
   Metadata md
}

Reply:
re: `ReservationTimedOut` 

I think if the reservation expired, the rest of the tasks should just fight for the spot. One task will get `ReservationAcquired` and the rest will get `AlreadyInProgress`